### PR TITLE
Add YOLOX object detection

### PR DESCRIPTION
**Pull request by laggui**
Closes issue #21 

This PR adds a Burn implementation for the YOLOX object detection models. Implementation and weights are based on [the official implementation](https://github.com/Megvii-BaseDetection/YOLOX).

**TODO**
- [x] YOLOX-Nano support
  - Requires [enum deserializer support](https://github.com/tracel-ai/burn/issues/1431) to switch between regular conv blocks and depthwise separable conv blocks
- [ ] ~NMS tensor implementation~
  - ~Requires `argsort`~
  - ~We should compare timings w/ the current naive implementation to see if it is worth it since we don't have the nms kernel to perform the computations even more efficiently~
 
 I scratched the NMS tensor implementation. Not required at this time.

### CODE REVIEW

File: yolox-burn/examples/inference.rs
Comment (antimora) [inline yolox-burn/examples/inference.rs:22]: We have permute now.

File: yolox-burn/src/model/head.rs
Comment (nathanielsimard) [inline yolox-burn/src/model/head.rs:127]: Dead comment?

**Review by nathanielsimard — COMMENTED** [review]
Some minor comments, but LGTM

File: yolox-burn/README.md
Comment (laggui) [reply yolox-burn/README.md:25]: Nope! copy-pasta error, the pre-trained weights are from COCO.

File: yolox-burn/src/model/head.rs
Comment (laggui) [reply yolox-burn/src/model/head.rs:127]: Yeah I initially thought about implementing the forward for inference and train (which requires other intermediate outputs) so I left this TODO comment. But I think we should just start with the inference support and implement training in another PR.

File: yolox-burn/src/model/blocks.rs
Comment (antimora) [inline yolox-burn/src/model/blocks.rs:23]: Some document strings will be help for variants here.

File: yolox-burn/src/model/blocks.rs
Comment (antimora) [inline yolox-burn/src/model/blocks.rs:110]: is this comment "// SiLU not in nn yet" still valid?

File: yolox-burn/src/model/weights.rs
Comment (antimora) [inline yolox-burn/src/model/weights.rs:65]: It would be great if we come up with a strategy/solution for the users to specify/override the URLs. I could imagine links could be broken/changed, or simply the user has its own back up repository. No need to fix it now, but we should be thinking this.

File: yolox-burn/src/model/weights.rs
Comment (antimora) [reply yolox-burn/src/model/weights.rs:131]: Ahh, that's fine to me - it seems super minor.

File: yolox-burn/src/model/yolox.rs
Comment (antimora) [inline yolox-burn/src/model/yolox.rs:19]: This TODO still valid?

File: yolox-burn/src/model/yolox.rs
Comment (antimora) [inline yolox-burn/src/model/yolox.rs:307]: Dead code?

File: yolox-burn/src/model/yolox.rs
Comment (antimora) [inline yolox-burn/src/model/yolox.rs:271]: It would be great also if we resave to Burn native format. PyTorch format is slower and has more dependency. 

Not required to fix this now but we should be thinking it doing it. We can enhance in a next PR.

**Review by antimora — APPROVED** [review]
Wow that's big. Looks good overall. I have minor comments. Please see. I am approving ahead.

File: yolox-burn/src/model/yolox.rs
Comment (laggui) [reply yolox-burn/src/model/yolox.rs:19]: Man I swear I always forget a dead comment somewhere 🤣 

The comment is not valid, I just added the Nano version with the recent support (thanks to you) for enums in deserialization.

File: yolox-burn/src/model/yolox.rs
Comment (laggui) [reply yolox-burn/src/model/yolox.rs:307]: Yup, another one 😅

File: yolox-burn/src/model/yolox.rs
Comment (laggui) [reply yolox-burn/src/model/yolox.rs:271]: Yeah we could do that for all the imported weights. We just need to store them somewhere, possibly on HF hub? We could keep the discussion going [here](https://github.com/tracel-ai/models/issues/16)

File: yolox-burn/src/model/blocks.rs
Comment (laggui) [reply yolox-burn/src/model/blocks.rs:110]: Silu is not implemented as a module like Relu or Gelu in burn. But I can definitely remove that comment, it doesn't serve much of a purpose.

File: yolox-burn/src/model/weights.rs
Comment (laggui) [reply yolox-burn/src/model/weights.rs:65]: I agree, it would be great to come up with a solution. Right now I just followed the same pattern as with ResNet, but we could finalize the discussion [here](https://github.com/tracel-ai/models/issues/16).

**Review by nathanielsimard — APPROVED** [review]

**Comment by antimora** [comment]
@laggui ready to merge?

**Comment by laggui** [comment]
> @laggui ready to merge?

Actually I think we're waiting for the Burn release so we can use the crates version and announce some things.

I'll let @nathanielsimard decide when we merge this.

**Review by nathanielsimard — APPROVED** [review]
🎉

**Pull request merged**
